packages/editor/src/optimizer/common.ts

```diff
@@ -0,0 +1,614 @@
+ */
+export type LayoutState = {
+  canvas: Canvas;
+  shapes: Shape<number>[];
+  labelMeasurements: LabelMeasurements;
+  variation: string;
+};
+
+/** Minimal state needed for rendering */
+export type RenderState = {
+  canvas: Canvas;
+  shapes: Shape<number>[];
+  labelCache: LabelCache;
+  variation: string;
+};
+
+/** Info for a successful compile */
+export type CompileInfo = {
+  diagramId: DiagramID;
+  warnings: PenroseWarning[];
+};
+
+// Message types
+
+/** Requests sent from main thread to broker or broker to worker */
+export type MessageRequest = {
+  tag: "MessageRequest";
+  messageId: MessageID;
+  data: MessageRequestData;
+};
+
+/** Responses to `MessageRequest`s */
+export type MessageResponse = {
+  tag: "MessageResponse";
+  messageId: MessageID;
+  result: MessageResult;
+};
+
+/** Unprompted notifications between threads. No response can be sent. */
+export type Notification = {
+  tag: "Notification";
+  data: NotificationData;
+};
+
+/** Data sent in a `MessageRequest` */
+export type MessageRequestData =
+  | CompileRequestData
+  | PollRequestData
+  | ComputeLayoutRequestData
+  | ResampleRequestData;
+
+/** Result type contained in a `MessageResponse` */
+export type MessageResult =
+  | CompileResult
+  | PollResult
+  | ComputeLayoutResult
+  | ResampleResult;
+
+/** Data contained in a `Notification` */
+export type NotificationData =
+  | InitData
+  | LabelMeasurementData
+  | DiscardDiagramData;
+
+/** Ensure request data and result pairs have the same tag, to check validity */
+export enum MessageTags {
+  Compile = "Compile",
+  Poll = "Poll",
+  ComputeLayout = "ComputeLayout",
+  DiscardDiagram = "DiscardDiagram",
+  DiscardStepSequence = "DiscardStepSequence",
+  Init = "Init",
+  LabelMeasurements = "LabelMeasurements",
+  Resample = "Resample",
+}
+
+// Request Data
+
+export type CompileRequestData = {
+  tag: MessageTags.Compile;
+  domain: string;
+  style: string;
+  substance: string;
+  variation: string;
+};
+
+export type PollRequestData = {
+  tag: MessageTags.Poll;
+  diagramId: number;
+};
+
+export type ComputeLayoutRequestData = {
+  tag: MessageTags.ComputeLayout;
+  diagramId: DiagramID;
+  historyLoc: HistoryLoc;
+};
+
+export type ResampleRequestData = {
+  tag: MessageTags.Resample;
+  diagramId: DiagramID;
+  variation: string;
+};
+
+// Notification Data
+
+export type InitData = {
+  tag: MessageTags.Init;
+};
+
+export type LabelMeasurementData = {
+  tag: MessageTags.LabelMeasurements;
+  diagramId: DiagramID;
+  labelMeasurements: LabelMeasurements;
+};
+
+export type DiscardDiagramData = {
+  tag: MessageTags.DiscardDiagram;
+  diagramId: DiagramID;
+};
+
+// Results
+
+export type TaggedOk<V, T> = {
+  tag: T;
+  variant: "Ok";
+  value: V;
+};
+
+export type TaggedErr<E, T> = {
+  tag: T;
+  variant: "Err";
+  error: E;
+};
+
+/**
+ * A Result type, but with a tag. Supertype of `MessageResult`.
+ */
+export type TaggedResult<V, E, T> = TaggedOk<V, T> | TaggedErr<E, T>;
```

Comment:
Do you still need these custom result types if you're using the `true-myth` library?

Reply:
Gotcha :/ yeah that makes sense; up to you if you want to get rid of `true-myth` entirely. I don't think getting rid of the dependency matters much per se.